For #17: created test for filelist creation option

paulodamaso opened the pull request:
For #17: created test for filelist creation option
- Created field and changed constructors in `Rpm` to accept filelist creation flag
- Added test which checks if `filelist.xml` and `filelist.xml.gz` files are created
- Left puzzle to implement the feature

0crat commented:
This pull request [#42](https://github.com/artipie/rpm-adapter/issues/42) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/3c70720c-8320-49c1-ba1e-534ed5cd9242); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

<!-- https://www.0crat.com/footprint/CT2E6TK9B/d54fb3db-238a-448c-965a-4b2084f2cabe, version: 0.51.12, hash: ${buildNumber} -->

paulodamaso commented:
@amihaiemil PLease review, seems that the failures aren't directly related to my changes

0crat commented:
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

<!-- https://www.0crat.com/footprint/CT2E6TK9B/cbca5fe9-f627-4694-a01f-e43f5fdd6e43, version: 0.51.13, hash: ${buildNumber} -->

amihaiemil commented:
@0crat refuse

0crat commented:
> @0crat refuse ([here](https://github.com/artipie/rpm-adapter/issues/42#issuecomment-602587065))

@amihaiemil The user @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) resigned from [#42](https://github.com/artipie/rpm-adapter/issues/42), please stop working. Reason for job resignation: Order was cancelled

<!-- https://www.0crat.com/footprint/CT2E6TK9B/a9ec15b1-7395-4bad-b5ce-3997a3d213e4, version: 0.51.13, hash: ${buildNumber} -->

0crat commented:
This pull request [#42](https://github.com/artipie/rpm-adapter/issues/42) is assigned to @vatavuk[/z](https://www.0crat.com/u/vatavuk), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/82f37d82-e02e-436d-af2e-29f4733d5ef1); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

<!-- https://www.0crat.com/footprint/CT2E6TK9B/82a4bbb9-e402-4bd6-8b31-fa9147b1ed15, version: 0.51.13, hash: ${buildNumber} -->

Vatavuk commented:
@paulodamaso you have conflicts

Vatavuk commented on src/test/java/com/artipie/rpm/RpmTest.java:
```diff
@@ -63,27 +67,62 @@
+            storage.list(new Key.From(RpmTest.REPODATA))
+                .get().stream().map(Key::string).collect(Collectors.toList()),
+            new IsIterableContainingInAnyOrder<>(
+                new ListOf<Matcher<? super String>>(
```
@paulodamaso we should assert that these files do not exist.

Vatavuk commented on src/main/java/com/artipie/rpm/Rpm.java:
```diff
@@ -81,25 +87,32 @@
+     * @param filelistsgen Flag to generate the filelist.
      */
-    public Rpm(final Storage stg, final NamingPolicy naming) {
+    public Rpm(final Storage stg, final NamingPolicy naming, final boolean filelistsgen) {
```
@paulodamaso Do we have an alternative to a boolean parameter? Using this class is not user friendly:
` new Rpm(storage, NamingPolicy.DEFAULT, false)` How about using `factory` classes or wrapping the boolean to an object?

paulodamaso commented:
@Vatavuk Thanks for the review. I've fixed the boolean flag problem throgh the use of decoration, please take a look. As for the test, due changes in introduced in #46 `Rpm` does not create the files anymore, so the test now is correct. I just moved it to the correct class, and it is still disabled because the behavior is not implemented yet.

paulodamaso commented:
@Vatavuk Ping

Vatavuk commented on src/main/java/com/artipie/rpm/WithFilelists.java:
```diff
@@ -0,0 +1,62 @@
+/**
+ * Rpm decorator which provide filelist creation on update.
+ *
+ * @since 0.1
```
@paulodamaso since 0.3.3

Vatavuk commented on src/test/java/com/artipie/rpm/WithFilelistsTest.java:
```diff
@@ -0,0 +1,86 @@
+/**
+ * Test for {@link WithFilelists}.
+ *
+ * @since 0.1
```
@paulodamaso since 0.3.3

paulodamaso commented:
@Vatavuk Done, please take a look

g4s8 commented:
@paulodamaso please resolve merge conflicts

paulodamaso commented:
@g4s8 Done!

Vatavuk commented:
@g4s8 ping

Vatavuk commented:
@paulodamaso confilcts

paulodamaso commented:
@g4s8 done

0crat commented:
@vatavuk[/z](https://www.0crat.com/u/vatavuk) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

<!-- https://www.0crat.com/footprint/CT2E6TK9B/6ed760ca-d202-4e06-9db3-25e596b22d84, version: 0.52.3, hash: ${buildNumber} -->

paulodamaso commented:
@Vatavuk Ping

Vatavuk commented:
@g4s8 we can merge

Vatavuk commented:
@g4s8 pls merge or refuse, I have only two days left to get paid.

paulodamaso commented:
@g4s8 please merge

g4s8 commented:
@Vatavuk 0crat won't resign you on PRs, 10 days rule works only for issues, not PRs

paulodamaso commented:
> @paulodamaso there are some concerns about this PR: RPM adapter is already in use by our clients, so we can't change its API completely, it leads to big refactoring on their side. Is it possible to hide this implementation under some additional constructor params, e.g. `new Rpm(storage, new FilelistsOptions(/*include: */: true))` or something similar and keep current secondary constructors as is without changes?

@g4s8 It was my first implementation, but me and @Vatavuk agreed that it wasn't correctly designed. What if I:
- implement this design I'm proposing with another naming
- wrap it with current `Rpm` to avoid client code refactor
- add a constructor to `Rpm` which would be used to create adapter with filelist option
- fix tests (obviously)

I think that this way we could at least keep a clean design in the API core and avoid client code refactoring, WDYT?

g4s8 commented:
> > @paulodamaso there are some concerns about this PR: RPM adapter is already in use by our clients, so we can't change its API completely, it leads to big refactoring on their side. Is it possible to hide this implementation under some additional constructor params, e.g. `new Rpm(storage, new FilelistsOptions(/*include: */: true))` or something similar and keep current secondary constructors as is without changes?
> 
> @g4s8 It was my first implementation, but me and @Vatavuk agreed that it wasn't correctly designed. What if I:
> 
>     * implement this design I'm proposing with another naming
> 
>     * wrap it with current `Rpm` to avoid client code refactor
> 
>     * add a constructor to `Rpm` which would be used to create adapter with filelist option
> 
>     * fix tests (obviously)
> 
> 
> I think that this way we could at least keep a clean design in the API core and avoid client code refactoring, WDYT?

@paulodamaso yes, it should work

paulodamaso commented:
@g4s8 Could you please boost it? I've remade this PR several times due to conflict and will have to make it again due this architectural change

paulodamaso commented:
@g4s8 Done, please take a look

g4s8 commented:
@paulodamaso please check build failures

paulodamaso commented:
@g4s8 Done, please take a look

0crat commented:
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @vatavuk[/z](https://www.0crat.com/u/vatavuk), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

<!-- https://www.0crat.com/footprint/CT2E6TK9B/18ed8520-35dc-4282-b3e2-50c3188cb9a7, version: 0.52.15, hash: ${buildNumber} -->

0crat commented:
Code review was too long (11 days), architects (@g4s8) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

<!-- https://www.0crat.com/footprint/CT2E6TK9B/4e3ae643-419f-42ed-93c6-04db6e2bb5a9, version: 0.52.15, hash: ${buildNumber} -->

sereshqua commented:
@0crat quality good

Sammers21 commented:
@paulodamaso , 480 lines of the diff...
